packages/node-core/src/indexer/dictionary/dictionary.service.ts

```diff
@@ -0,0 +1,144 @@
+import {NodeConfig} from '../../configure';
+import {getLogger} from '../../logger';
+import {BlockHeightMap} from '../../utils/blockHeightMap';
+import {DictionaryResponse, IDictionary, IDictionaryCtrl} from './types';
+
+const logger = getLogger('DictionaryService');
+export abstract class DictionaryService<DS, FB> implements IDictionaryCtrl<DS, FB>, OnApplicationShutdown {
+  protected _dictionaries: IDictionary<DS, FB>[] = [];
+
+  protected _currentDictionaryIndex: number | undefined;
+  constructor(
+    protected chainId: string,
+    protected readonly nodeConfig: NodeConfig,
+    protected readonly eventEmitter: EventEmitter2
+  ) {}
+
+  onApplicationShutdown(): void {
+    this._dictionaries = [];
+    this._currentDictionaryIndex = undefined;
+  }
+
+  abstract initDictionaries(): Promise<void>;
+
+  init(dictionaries: IDictionary<DS, FB>[]): void {
+    this._dictionaries = dictionaries;
+  }
+
+  private getDictionary(height: number, skipDictionaryIndex: Set<number> = new Set()): IDictionary<DS, FB> | undefined {
+    if (this._dictionaries.length === 0) {
+      logger.debug(`No dictionaries available to use`);
+      return undefined;
+    }
+    // If current dictionary is valid, use current one instead of find a dictionary
+    if (
+      this._currentDictionaryIndex !== undefined &&
+      !skipDictionaryIndex.has(this._currentDictionaryIndex) &&
+      this._dictionaries[this._currentDictionaryIndex].heightValidation(height)
+    ) {
+      return this._dictionaries[this._currentDictionaryIndex];
+    } else {
+      this.findDictionary(height, skipDictionaryIndex);
+      if (this._currentDictionaryIndex === undefined) {
+        logger.warn(`No supported dictionary found`);
+        return undefined;
+      } else {
+        logger.debug(`Updated : current dictionary Index is ${this._currentDictionaryIndex}`);
+        return this._dictionaries[this._currentDictionaryIndex];
+      }
+    }
+  }
+
+  // Find the next valid dictionary
+  private findDictionary(height: number, skipDictionaryIndex: Set<number>) {
+    // DO NOT remove dictionary not valid
+    // As they can be valid for different block range, or work for other query
+    const index = this._dictionaries.findIndex(
+      (d, index) => d.heightValidation(height) && !skipDictionaryIndex.has(index)
+    );
+    // If not find any, then still set undefined.
+    this._currentDictionaryIndex = index < 0 ? undefined : index;
+  }
+
+  useDictionary(height: number): boolean {
+    return !!this.getDictionary(height);
+  }
+
+  /**
+   *
+   * @param dataSources
+   */
+
+  buildDictionaryEntryMap(dataSources: BlockHeightMap<DS[]>): void {
+    for (const dict of this._dictionaries) {
+      dict.updateQueriesMap(dataSources);
+    }
+  }
+
+  async scopedDictionaryEntries(
+    startBlockHeight: number,
+    scaledBatchSize: number,
+    latestFinalizedHeight: number, //api FinalizedHeight
+    skipDictionaryIndex: Set<number> = new Set<number>()
```

Comment:
I think this function should be broken up with a private version that hides this argument